elyra/pipeline/registry.py

```diff
@@ -0,0 +1,623 @@
+
+        # Append input group info to parameter details
+        component_parameters['uihints']['group_info'][0]['group_info'].append(input_group_info)
+
+        # Define new output group object
+        output_group_info = {
+            'id': "outputs",  # need to actually figure out the control id
+            'type': "controls",
+            'parameter_refs': []
+        }
+
+        outputs = component_body['outputs']
+        for output_object in outputs:
+            new_parameter_info = self.build_parameter(output_object, "output")
+
+            if new_parameter_info['parameter_ref'] in input_group_info['parameter_refs']:
+                new_parameter_info['parameter_ref'] = f"output_{new_parameter_info['parameter_ref']}"
+
+            # Add to existing parameter list
+            component_parameters['parameters'].append({"id": new_parameter_info['parameter_ref']})
+            component_parameters['current_parameters'][new_parameter_info['parameter_ref']] = ""
+
+            # Add to existing parameter info list
+            component_parameters['uihints']['parameter_info'].append(new_parameter_info)
+
+            # Add parameter to output group info
+            output_group_info['parameter_refs'].append(new_parameter_info['parameter_ref'])
+
+        # Append output group info to parameter details
+        component_parameters['uihints']['group_info'][0]['group_info'].append(output_group_info)
+
+        return component_parameters
+
+    def build_parameter(self, obj, obj_type):
+        data_object = {}
+        # Determine whether parameter is optional
+        if ("optional" in obj and obj['optional']) \
+                or ("description" in obj and "required" in obj['description'].lower()):
+            data_object['required'] = True
+        else:
+            data_object['required'] = False
+
+        # Assign type, default to string
+        data_object['format'] = "string"
+        custom_control_id = "StringControl"
+        if "type" in obj:
+            data_object['format'] = obj['type']
+
+            # This may not be applicable in every case
+            if "string" in obj['type'].lower():
+                custom_control_id = "StringControl"
+            elif obj['type'].lower() in ["number", "integer"]:
+                custom_control_id = "NumberControl"
+            elif obj['type'].lower() in ["bool", "boolean"]:
+                custom_control_id = "BooleanControl"
+            elif "array" in obj['type'].lower():
+                custom_control_id = "StringArrayControl"
+            elif "object" in obj['type'].lower():
+                custom_control_id = "StringControl"
+
+        # Build label name
+        label = obj['name']
+        if obj_type not in obj['name'].lower():
+            label = f"{obj['name']} ({obj_type})"
+
+        # Build parameter info
+        new_parameter = {
+            'parameter_ref': obj['name'].replace(' ', '_'),
```

Comment:
Since the parameter structure is identical across ComponentParsers, it might be better to add a method on the superclass that takes the 5 parameters that are runtime-specific and returns the parameter JSON. 

The same comment goes for determining `custom_control_id`.  That would also prevent the redundant calls to `lower()`.

Reply:
Yeah, the more `lower()` calls I added, I started to question myself 😄 . Great idea!